### PR TITLE
[AIRFLOW-3995][WIP] Extract code from docs

### PR DESCRIPTION
**Pull request by mik-laj**
Make sure you have checked _all_ steps below.

### Jira

  - https://issues.apache.org/jira/browse/AIRFLOW-3995

### Description

I extracted fragments of codes to separate files. This allows us to test them automatically. Some examples were incorrect, so I fixed them.

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [ ] Passes `flake8`

### CODE REVIEW

File: tests/plugins/test_plugin.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
 # specific language governing permissions and limitations
 # under the License.
 
+# [START plugins_interface]
```
Comment (potiuk) [inline tests/plugins/test_plugin.py]: It should be "plugin_example"

File: tests/test_documentation.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# -*- coding: utf-8 -*-
```
Comment (potiuk) [inline tests/test_documentation.py]: Nice. We are testing the docs :).

File: tests/www/test_utils.py [hunk 3]
```diff
@@ -16,13 +16,26 @@
...
+from datetime import datetime
+
 import mock
 from xml.dom import minidom
 
+from airflow import models, AirflowException
 from airflow.www import utils
+from airflow.www.utils import render_attrs
+
+AIRFLOW_MAIN_FOLDER = os.path.realpath(
```
Comment (potiuk) [inline tests/www/test_utils.py]: I actually found a simpler way in airflow code: airflow.example_dags.__path__[0] - maybe that approach can be used here and in other places.

File: tests/www/test_utils.py [hunk 4]
```diff
@@ -148,6 +161,34 @@ def test_open_maybe_zipped_archive(self, mocked_ZipFile, mocked_is_zipfile):
         (args, kwargs) = instance.open.call_args_list[0]
         self.assertEqual('deep/path/to/file.txt', args[0])
 
+    def test_render_example_dagsself(self):
```
Comment (potiuk) [inline tests/www/test_utils.py]: Really nice. Maybe we should do the same for ALL example dags in a loop ! That would be awesome if we could do that!

File: airflow/example_dags/example_bash_operator.py [hunk 5]
```diff
@@ -69,5 +69,16 @@
 # [END howto_operator_bash_template]
 also_run_this >> run_this_last
 
+
+# [END concepts_jinja_templating]
```
[the remark this answers is not in the document]
Comment (mik-laj) [reply airflow/example_dags/example_bash_operator.py]: Fixed.

File: docs/lineage.rst [hunk 6]
```diff
@@ -26,42 +26,10 @@ audit trails and data governance, but also debugging of data flows.
...
-        outlets.append(f_out)
-    run_this = BashOperator(    
-        task_id='run_me_first', bash_command='echo 1', dag=dag,
-        inlets={"datasets": [f_in,]},
-        outlets={"datasets": outlets}
-        )
-    run_this.set_downstream(run_this_last)
+.. literalinclude:: ../airflow/example_dags/example_lineage.py
+    :language: python
+    :start-after: [START lineage]
```
[the remark this answers is not in the document]
Comment (mik-laj) [reply docs/lineage.rst]: Fixed. Thanks.

File: tests/plugins/test_plugin.py (hunk 1 above)
Comment (mik-laj) [reply tests/plugins/test_plugin.py]: I use conventions "[file name] _ [section name]". This allows us to easily find the place where the fragment is loaded.

File: tests/www/test_utils.py (hunk 3 above)
Comment (mik-laj) [reply tests/www/test_utils.py]: Thanks. I apply your suggestion.

File: tests/www/test_utils.py (hunk 4 above)
Comment (mik-laj) [reply tests/www/test_utils.py]: A good idea. I added this solution ..

File: tests/test_documentation.py (hunk 2 above)
Comment (mik-laj) [reply tests/test_documentation.py]: Next level :dart:

File: docs/concepts.rst [hunk 7]
```diff
@@ -887,20 +887,11 @@ This is especially useful if your tasks are built dynamically from
...
-
-    """
-    ### My great DAG
-    """
-
-    dag = DAG('my_dag', default_args=default_args)
-    dag.doc_md = __doc__
+.. literalinclude:: ../airflow/example_dags/example_documentation.py
+    :language: python
+    :start-after: [START concepts_documentation]
```
[the remark this answers is not in the document]
Comment (mik-laj) [reply docs/concepts.rst]: Fixed.

**Comment by mik-laj** [comment]
To reduce the number of conflicts, I will regularly move your commit to my branch.

**Pull request closed without merging**